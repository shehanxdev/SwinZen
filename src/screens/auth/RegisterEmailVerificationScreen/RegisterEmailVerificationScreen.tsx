import { yupResolver } from '@hookform/resolvers/yup';
import React from 'react';
import { Controller, SubmitErrorHandler, SubmitHandler, useForm } from 'react-hook-form';
import { View } from 'react-native';

import { Button, Link, SwingZenLogoIcon, Text } from '@sz/components';
import { tw } from '@sz/config';
import { OtpType, Route, TextVariant } from '@sz/constants';
import { OtpVerficationValue } from '@sz/models';
import { NavigationService, ToastService } from '@sz/services';
import { useDispatch, useSelector } from '@sz/stores';
import { getMaskedMail, otpValidationSchema } from '@sz/utils';

import { BaseAuthScreen, OTPInput } from '../components';
import { ResendOtpWithTimer } from '../components/ResendOtpWithTimer';

export function RegisterEmailVerificationScreen({ route }) {
  const {
    control,
    handleSubmit,
    formState: { errors },
    getValues,
  } = useForm<OtpVerficationValue>({ mode: 'onChange', resolver: yupResolver(otpValidationSchema) });

  const username = route.params.params;

  const loading = useSelector(state => state.loading.effects.userStore.emailVerification);

  const dispatch = useDispatch();

  const onRegisterEmailFormInvalid: SubmitErrorHandler<OtpVerficationValue> = () => {
    console.log(errors);
    //TODO:: handle error
  };

  const onRegisterEmailFormValid: SubmitHandler<OtpVerficationValue> = () => {
    console.log('success');
  };

  const onResend = async () => {
    try {
      await dispatch.userStore.resendOtp({ username: username });
      ToastService.success({ message: 'Success!', description: 'OTP resent successfully.' });
    } catch (error: any) {
      ToastService.error({ message: 'Failed!', description: error.data.message });
    }
  };

  const onVerify = async () => {
    const otpData = {
      username: username,
      otpType: OtpType.VERIFICATION,
      otp: getValues('otp'),
    };
    try {
      await dispatch.userStore.emailVerification(otpData);
      NavigationService.navigate(Route.Login);
    } catch (error: any) {
      ToastService.error({ message: 'Failed!', description: error.data.message });
    }
  };

  return (
    <BaseAuthScreen>
      <View style={tw`flex-1 justify-between`} testID="RegisterEmailVerificationScreenContainerTestID">
        <View style={tw`mt-8.75 mx-4`}>
          <View style={tw`items-center`}>
            <SwingZenLogoIcon />
          </View>
          <View style={tw`items-center`}>
            <View style={tw`mt-6 mb-2`}>
              <Text variant={TextVariant.SubTitle2SemiBold}>Email verification</Text>
            </View>
            <View style={tw`mb-10`}>
              <Text variant={TextVariant.Body2Regular}>
                {`Enter the code received in your email address ${getMaskedMail(username)}`}
              </Text>
            </View>
          </View>
          <Controller
            control={control}
            name="otp"
            render={({ field: { value, onChange } }) => (
              <OTPInput
                value={value}
                onChangeValue={onChange}
                onSubmitEditing={handleSubmit(onRegisterEmailFormValid, onRegisterEmailFormInvalid)}
              />
            )}
          />
          <View style={tw`items-center mt-6`}>
            <ResendOtpWithTimer onResend={onResend} />
          </View>
        </View>
        <View style={tw`items-center mb-5 items-center mx-5 mt-15`}>
          <View style={tw`mb-6`}>
            <Button onPress={onVerify} title="verify" loading={loading} />
          </View>
          <Text variant={TextVariant.Body2Regular}>
            By continuing, you agree to our{' '}
            <Link
              text="Privacy Policy"
              onPress={() => {
                NavigationService.navigate(Route.PrivacyPolicy);
              }}
            />{' '}
            and our{' '}
            <Link
              text="Terms of Use"
              onPress={() => {
                NavigationService.navigate(Route.TermsOfUse);
              }}
            />
          </Text>
        </View>
      </View>
    </BaseAuthScreen>
  );
}
