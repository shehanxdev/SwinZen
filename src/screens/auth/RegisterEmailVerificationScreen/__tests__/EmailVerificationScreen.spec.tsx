import { render } from '@testing-library/react-native';
import * as React from 'react';

import { RegisterEmailVerificationScreen } from '../RegisterEmailVerificationScreen';

describe('Register Email Verification Screen', () => {
  const testID = 'RegisterEmailVerificationScreenContainerTestID';
  const getRenderedScreen = () => render(<RegisterEmailVerificationScreen />);

  it(`should render correctly`, () => {
    const rendered = getRenderedScreen();
    const renderedTree = rendered.toJSON();
    expect(renderedTree).toMatchSnapshot();
  });

  it('should find the screen via testID', () => {
    const { getByTestId } = getRenderedScreen();
    const foundScreen = getByTestId(testID);

    expect(foundScreen).toBeTruthy();
  });
});
