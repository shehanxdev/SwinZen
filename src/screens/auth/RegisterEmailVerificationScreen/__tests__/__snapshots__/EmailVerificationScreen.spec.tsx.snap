// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Register Email Verification Screen should render correctly 1`] = `
<BVLinearGradient
  colors={
    [
      4279320343,
      4278650887,
    ]
  }
  endPoint={
    {
      "x": 0.5,
      "y": 1,
    }
  }
  locations={null}
  startPoint={
    {
      "x": 0.5,
      "y": 0,
    }
  }
  style={
    {
      "flexBasis": "0%",
      "flexGrow": 1,
      "flexShrink": 1,
    }
  }
>
  <View
    style={
      {
        "flexBasis": "0%",
        "flexGrow": 1,
        "flexShrink": 1,
        "justifyContent": "space-between",
      }
    }
    testID="RegisterEmailVerificationScreenContainerTestID"
  >
    <View
      style={
        {
          "marginLeft": 20,
          "marginRight": 20,
          "marginTop": 80,
        }
      }
    >
      <View
        style={
          {
            "alignItems": "center",
          }
        }
      >
        <Image
          source={
            {
              "testUri": "../../../src/assets/logo/swingzen-logo.png",
            }
          }
        />
      </View>
      <View
        style={
          {
            "alignItems": "center",
          }
        }
      >
        <View
          style={
            {
              "marginBottom": 90,
              "marginTop": 12,
            }
          }
        >
          <Text
            style={
              [
                {
                  "color": "#F8F9FA",
                  "lineHeight": 28,
                },
                {
                  "fontFamily": "SourceSansPro-SemiBold",
                  "fontSize": 21,
                  "fontWeight": "600",
                  "textAlign": "center",
                },
              ]
            }
          >
            Email verification
          </Text>
        </View>
        <View
          style={
            {
              "marginBottom": 24,
            }
          }
        >
          <Text
            style={
              [
                {
                  "color": "#F8F9FA",
                  "lineHeight": 24,
                },
                {
                  "fontFamily": "SourceSansPro-Regular",
                  "fontSize": 16,
                  "fontWeight": "normal",
                  "textAlign": "center",
                },
              ]
            }
          >
            Enter the code received in your email address s**t**z@gmail.com
          </Text>
        </View>
      </View>
      <View
        style={
          [
            {
              "flexDirection": "row",
              "justifyContent": "space-between",
            },
            {
              "height": 72,
            },
          ]
        }
      >
        <View
          onLayout={[Function]}
          style={
            [
              {
                "alignItems": "center",
                "borderRadius": 10,
                "height": "100%",
                "justifyContent": "center",
                "width": 58,
              },
              {
                "backgroundColor": "#283E36",
              },
            ]
          }
        >
          <Text
            style={
              [
                {
                  "color": "#F8F9FA",
                  "lineHeight": 28,
                },
                {
                  "fontFamily": "SourceSansPro-SemiBold",
                  "fontSize": 21,
                  "fontWeight": "600",
                  "textAlign": "center",
                },
              ]
            }
          />
          <View
            style={
              {
                "backgroundColor": "#6C757D",
                "bottom": 17,
                "height": 1,
                "position": "absolute",
                "width": 20,
              }
            }
          />
        </View>
        <View
          onLayout={[Function]}
          style={
            [
              {
                "alignItems": "center",
                "borderRadius": 10,
                "height": "100%",
                "justifyContent": "center",
                "width": 58,
              },
              {
                "backgroundColor": "#283E36",
              },
            ]
          }
        >
          <Text
            style={
              [
                {
                  "color": "#F8F9FA",
                  "lineHeight": 28,
                },
                {
                  "fontFamily": "SourceSansPro-SemiBold",
                  "fontSize": 21,
                  "fontWeight": "600",
                  "textAlign": "center",
                },
              ]
            }
          />
          <View
            style={
              {
                "backgroundColor": "#6C757D",
                "bottom": 17,
                "height": 1,
                "position": "absolute",
                "width": 20,
              }
            }
          />
        </View>
        <View
          onLayout={[Function]}
          style={
            [
              {
                "alignItems": "center",
                "borderRadius": 10,
                "height": "100%",
                "justifyContent": "center",
                "width": 58,
              },
              {
                "backgroundColor": "#283E36",
              },
            ]
          }
        >
          <Text
            style={
              [
                {
                  "color": "#F8F9FA",
                  "lineHeight": 28,
                },
                {
                  "fontFamily": "SourceSansPro-SemiBold",
                  "fontSize": 21,
                  "fontWeight": "600",
                  "textAlign": "center",
                },
              ]
            }
          />
          <View
            style={
              {
                "backgroundColor": "#6C757D",
                "bottom": 17,
                "height": 1,
                "position": "absolute",
                "width": 20,
              }
            }
          />
        </View>
        <View
          onLayout={[Function]}
          style={
            [
              {
                "alignItems": "center",
                "borderRadius": 10,
                "height": "100%",
                "justifyContent": "center",
                "width": 58,
              },
              {
                "backgroundColor": "#283E36",
              },
            ]
          }
        >
          <Text
            style={
              [
                {
                  "color": "#F8F9FA",
                  "lineHeight": 28,
                },
                {
                  "fontFamily": "SourceSansPro-SemiBold",
                  "fontSize": 21,
                  "fontWeight": "600",
                  "textAlign": "center",
                },
              ]
            }
          />
          <View
            style={
              {
                "backgroundColor": "#6C757D",
                "bottom": 17,
                "height": 1,
                "position": "absolute",
                "width": 20,
              }
            }
          />
        </View>
        <View
          onLayout={[Function]}
          style={
            [
              {
                "alignItems": "center",
                "borderRadius": 10,
                "height": "100%",
                "justifyContent": "center",
                "width": 58,
              },
              {
                "backgroundColor": "#283E36",
              },
            ]
          }
        >
          <Text
            style={
              [
                {
                  "color": "#F8F9FA",
                  "lineHeight": 28,
                },
                {
                  "fontFamily": "SourceSansPro-SemiBold",
                  "fontSize": 21,
                  "fontWeight": "600",
                  "textAlign": "center",
                },
              ]
            }
          />
          <View
            style={
              {
                "backgroundColor": "#6C757D",
                "bottom": 17,
                "height": 1,
                "position": "absolute",
                "width": 20,
              }
            }
          />
        </View>
        <TextInput
          autoCapitalize="characters"
          autoCorrect={false}
          blurOnSubmit={false}
          caretHidden={false}
          clearButtonMode="never"
          disableFullscreenUI={true}
          keyboardType="number-pad"
          maxLength={5}
          onBlur={[Function]}
          onChangeText={[Function]}
          onFocus={[Function]}
          onPressOut={[Function]}
          spellCheck={false}
          style={
            {
              "bottom": 0,
              "fontSize": 1,
              "left": 0,
              "opacity": 0.01,
              "position": "absolute",
              "right": 0,
              "top": 0,
            }
          }
          textContentType="oneTimeCode"
          underlineColorAndroid="transparent"
          value=""
        />
      </View>
      <View
        style={
          {
            "marginBottom": 12,
            "marginTop": 120,
          }
        }
      >
        <View
          style={
            {
              "flexDirection": "row",
            }
          }
        >
          <View
            collapsable={false}
            onTouchEnd={[Function]}
            onTouchStart={[Function]}
            style={
              {
                "backgroundColor": "#69AA67",
                "borderColor": "#ffffff00",
                "borderRadius": 10,
                "borderStyle": "solid",
                "borderWidth": 1,
                "elevation": 0,
                "justifyContent": "center",
                "minWidth": "auto",
                "shadowColor": "#000",
                "shadowOffset": {
                  "height": 0,
                  "width": 0,
                },
                "shadowOpacity": 0,
                "shadowRadius": 0,
                "width": "100%",
              }
            }
          >
            <View
              accessibilityRole="button"
              accessibilityState={
                {
                  "disabled": false,
                }
              }
              accessibilityValue={
                {
                  "max": undefined,
                  "min": undefined,
                  "now": undefined,
                  "text": undefined,
                }
              }
              accessible={true}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                [
                  {
                    "overflow": "hidden",
                  },
                  {
                    "borderRadius": 10,
                  },
                ]
              }
            >
              <View
                style={
                  [
                    {
                      "alignItems": "center",
                      "flexDirection": "row",
                      "justifyContent": "center",
                    },
                    {
                      "height": 48,
                    },
                  ]
                }
              >
                <Text
                  numberOfLines={1}
                  selectable={false}
                  style={
                    [
                      {
                        "color": "#000000",
                        "fontFamily": "System",
                        "fontWeight": "400",
                      },
                      {
                        "textAlign": "left",
                      },
                      [
                        {
                          "letterSpacing": 1,
                          "marginHorizontal": 16,
                          "marginVertical": 9,
                          "textAlign": "center",
                        },
                        {
                          "marginHorizontal": 8,
                        },
                        false,
                        {
                          "color": "#6200ee",
                          "fontFamily": "System",
                          "fontWeight": "500",
                        },
                        {
                          "fontFamily": "System",
                          "fontWeight": "500",
                        },
                        undefined,
                      ],
                    ]
                  }
                >
                  <Text
                    style={
                      [
                        {
                          "color": "#F8F9FA",
                          "lineHeight": 24,
                        },
                        {
                          "fontFamily": "SourceSansPro-SemiBold",
                          "fontSize": 16,
                          "fontWeight": "600",
                          "textAlign": "center",
                        },
                      ]
                    }
                  >
                    Verify
                  </Text>
                </Text>
              </View>
            </View>
          </View>
        </View>
      </View>
      <View
        style={
          {
            "alignItems": "center",
          }
        }
      >
        <Text
          style={
            [
              {
                "color": "#69AA67",
                "lineHeight": 20,
              },
              {
                "fontFamily": "SourceSansPro-Regular",
                "fontSize": 16,
                "fontWeight": "bold",
                "textAlign": "center",
              },
            ]
          }
        >
          Resend the code
        </Text>
      </View>
    </View>
    <View
      style={
        {
          "alignItems": "center",
          "marginBottom": 20,
          "marginLeft": 20,
          "marginRight": 20,
        }
      }
    >
      <Text
        style={
          [
            {
              "color": "#F8F9FA",
              "lineHeight": 24,
            },
            {
              "fontFamily": "SourceSansPro-Regular",
              "fontSize": 16,
              "fontWeight": "normal",
              "textAlign": "center",
            },
          ]
        }
      >
        By continuing, you agree to our
         
        <Text
          onPress={[Function]}
          style={
            [
              {
                "color": "#69AA67",
                "lineHeight": 20,
              },
              {
                "fontFamily": "SourceSansPro-Regular",
                "fontSize": 16,
                "fontWeight": "bold",
                "textAlign": "center",
              },
            ]
          }
        >
          Privacy Policy
        </Text>
         
        and our
         
        <Text
          onPress={[Function]}
          style={
            [
              {
                "color": "#69AA67",
                "lineHeight": 20,
              },
              {
                "fontFamily": "SourceSansPro-Regular",
                "fontSize": 16,
                "fontWeight": "bold",
                "textAlign": "center",
              },
            ]
          }
        >
          Terms of Use
        </Text>
      </Text>
    </View>
  </View>
</BVLinearGradient>
`;
